runtime/buffer.h

```diff
@@ -125,6 +128,18 @@ class raw_buffer : public ref_counted<raw_buffer> {
   void operator=(raw_buffer&&) = delete;
   ~raw_buffer() { free(); }
 
+  // This ctor is public, but only intended for use with the ALLOC_RAW_BUFFER_ON_STACK macro.
+  inline raw_buffer(
+      char* allocation_in, void* base_in, std::size_t elem_size_in, std::size_t rank_in, slinky::dim* dims_in, dim_init_fn d_fn)
```

Comment:
I don't think we can use std::function like this here. I doubt the compiler can "see" across that and inline appropriately. Even if it can, I don't like trusting the compiler to be so magically good. In most code in slinky I think it's fine to favor code style over performance, but not in `evaluate`.

I think a macro that allocates a buffer of a given rank on the stack is OK. I think if you just don't try to initialize dimensions in it, this gets a lot less tricky. Just repeating that code (initializing the dimensions) a few times seems fine to me. I also think we shouldn't have the macro in the buffer header, just in `evaluate.cc`.

Reply:
I didn't expect it to inline the code, but I also don't think the overhead of a lambda here really matters. that said, I'll take the partial fix, so I'll revise to leave the init of dims code inline.